mount/mount_windows.go

```diff
@@ -166,6 +185,18 @@ func removeSMBMapping(remotepath string) (string, error) {
 func (mounter *Mounter) Unmount(target string) error {
 	klog.V(4).Infof("azureMount: Unmount target (%q)", target)
 	target = NormalizeWindowsPath(target)
+	source, err := os.Readlink(target)
+	if err != nil {
+		klog.Infof("failed to get source from target %q, %v", target, err)
+		return err
+	}
+	output, err := exec.Command("powershell", "/c", "net use", source, "/delete").CombinedOutput()
```

Comment:
is this function used for unmount volume types other than nfs?

Reply:
currently only nfs and smb are supported here. For smb, it is not needed, but since we don't have fstype information during unmount, we cannot make different choices here. I tried net use source /delete, it does not return error if source does not exist. 

So unlike Linux, the mount/unmount functions used for different file systems in Windows are different.  We need to consider add fstype information in Unmount APIs, I think,

Reply:
actually, I checked this part again, and decided to remove this /delete code. Both smb and nfs have the same issue here. During mount, there are two steps
1. Set up network connection to remote share

- smb: this uses New-SmbGlobalMapping to set up the connection to the smb share

- nfs: use "net use" to set up the connection.

2. create a symlink from the remote connection to local path

During unmount, we will remove the symlink, but we could not delete the remote connection. If there are multiple pods on the same node share the same remote connection, deleting the remote connection will affect other pods. Or even outside of kubernetes, user might directly connect remote share to local host.

So currently, for smb, we do not remove the mapping during unmount, similarly, the option for nfs is to not delete the connection.

Reply:
When does it get cleaned up then?

Reply:
What happens if you mount the same share multiple times on the same node?

Reply:
It will not get cleaned up (delete connection)  because we cannot know when it is safe to do so. In linux, the mount is 1:1 mapping from remote to individual local path. So we could safely do unmount for each "mount". But in windows, remote connection is first connected to local host as a global connection, and each mount will create symlink from this global connection to individual path. 
If mount same share multiple times on the same node, it should work as expected. Do you see any issues with it?

Reply:
What happens if you create a connection and one already exists?

I'm concerned about resource leakage and maybe over time after many mounts, we will run out of resources on the node and will be unable to mount anymore. Or we will reach the max number of connections to an nfs/smb server due to all the stale client connections that still remain.  It seems like maybe we need to extend the nfs + smb plugins to support MountDevice so that it can clean up these resources.

Reply:
I could not find whether there is a limit on it. Will check with sig-windows. @andyzhangx 
I think we could extend plugin to support MountDevice, only concern here is out of band connection, and we might accidentally disconnect user's remote share.

Reply:
the main diff on smb mount between linux and windows is that on LInux, every smb mount is independent, while on Windows, one remote smb server address(e.g. `smb-server/fileshare`) can only have one SMB-GlobalMapping. So `if you create a connection and one already exists` on Windows, that is a difficulty:
 - in some cases, we should skip create SMB-GlobalMapping since it already exists
 - in other cases, original password is incorrect or expired, user wants to delete SMB-GlobalMapping, and then create a new SMB-GlobalMapping

On Windows, there is no way to check whether original SMB-GlobalMapping is valid or not, that was the answer about 1 years ago from windows team, will update this thread(also about the max SMB-GlobalMapping limit) if there is any new update recently.

Reply:
What is the behavior that we do now in the plugin? Is it possible to have the plugin do reference counting so that we can clean up the global mapping when all mounts are gone? We have the MountDevice interface which is supposed to automatically do the reference counting, but I'm not sure if it handles subshares correctly, or the same share referenced in multiple PVs.

Reply:
right now for smb global mapping, during setup, it will remove existing global mapping first due to the reason @andyzhangx mentioned above. There might cause issues for existing pod which uses the global mapping, but we haven't got a better way to handle it yet. https://github.com/kubernetes/utils/blob/f32512634ab7ee0599334344e026b1b47d39e4f4/mount/mount_windows.go#L103
During teardown, we do not do anything about the global mapping (no cleanup) except removing the symlink of local path from the mapping.

For nfs here, I didn't delete existing nfs connection during set up. We could execute "net use /nfsmountpoint" multiple times. We could do similar thing as smb, but both has issues. We could do some further discussion about it, and document the implication.

Reply:
For the smb case, does that mean we can't support multiple pods using the same share?

For the nfs case, what happens if you run "net use" multiple times on the same mount? Is there a chance we will exhaust resources on the node if we don't clean them up?

Reply:
for smb, when new smb mapping is created, I am not sure whether data access can be recovered or not. @andyzhangx do you have tests  for it with azure? I can also try csi smb driver later.

If run "net use" multiple times on the same mount, it will succeed. I don't see any additional mount is created, it will be the same one.

Reply:
@msau42 @jingxu97 that's a great discussion, I filed an issue here: https://github.com/kubernetes/kubernetes/issues/93805
will spend some time fixing the smb mapping issue.

Reply:
and about the SMBGlobalMapping limit, here is the answer from windows expert, so currently it's ok.
```
They're isn't any hard limit. But each global mapping will consume resources in the rdr. (Connections, authenticated sessions, file objects etc).  I'm assuming you are using a driveless global mapping. 
```

Reply:
there is a PR to fix the new smb mapping issue: https://github.com/kubernetes/utils/pull/184